### PR TITLE
codec/dagcbor: add DecodeOptions.ExperimentalDeterminism

### PR DESCRIPTION
**Pull request by mvdan**
(see commit message)

### CODE REVIEW

**Comment by mvdan** [comment]
We did apply https://github.com/ipld/go-ipld-prime/pull/386 by default and forcefully, but I imagine that one is extremely unlikely to break any users, whereas this one isn't.

**Comment by warpfork** [comment]
Yeah, I don't think we could do this by default.  Strict modes available when used as a library: good; strict modes being the default, esp being what happens when e.g. looking up a thing by multicodec indicator code: nonviable.

There's way too much data in the wild with noncanonical orders of map keys.  (The genesis block of filecoin is one of them... not that that particular example should have overwhelming weight, but it's certainly amusing.)

Noncanonical order is also extra-likely-in-practice for DAG-CBOR just because of the bizarreness of the CBOR RFC's choice of sorting algorithms; it catches people off-guard all the time.

**Comment by warpfork** [comment]
(edited previous comment, to make it more legible as general "case law" for future reference.)

**Comment by mvdan** [comment]
FWIW I fully expected this default to be rejected, hence the failing tests and the "DO NOT MERGE".

Tomorrow I'll update the PR to keep the strict mode opt-in on decoding, and update the fuzzer to not require the codecs to follow their own specs by default.

One open question is what should the option be: "enforce spec strictness on decode" to do everything at once (map key order,  canonical form for floats and integers, etc), or one option for each one of those separately, just like `MapSortMode`? I lean towards the former, because I can't imagine anyone would want a decoder that is "a little stricter but not fully strict". cc @rvagg

**Comment by mvdan** [comment]
Also, I know we've talked about this before, but: I think it's time to reconsider what the specs dictate. It's not good to have the dag-cbor core spec dictate strictness on decode when all current implementations do the opposite. The spec should be updated to reflect reality, given that it's not possible to update the defaults to reflect the spec.

**Comment by BigLep** [comment]
@mvdan : will create the spec update PR?

**Comment by rvagg** [comment]
2c:

* I pushed toward strictness _everywhere_ for a while but we ended up at this static state of accepting that there's just too much data in the wild that doesn't conform, as Eric outlined (this is even true of dag-pb, we can't even be completely strict with that!).
* The spec might lean toward suggesting strictness should be the default because it might be reflecting the directionality I was working toward and might need backing off a bit to reflect reality.
* I still would like an opt-in strictness in all our codecs--there will be systems that people build that are isolated enough that they _can_ enforce strictness and get the benefits from that (there's tradeoffs of course). I worked toward that in the JS codecs too, they're about as close as the Go ones to being able to enable some kind of switch.
* The approach I took in JS re options was to put a verbose suite of options in the base codec handling code ([cborg](https://github.com/rvagg/cborg)) but then take strong opinions in the interface for our IPLD codecs (see https://github.com/ipld/js-dag-cbor/blob/b34bf67543c91975d0afe051e50fa1d6eb66a35a/index.js#L86-L98) and I think I'd go a similar direction for JS if we exposed a strict-mode of some kind, just one switch that turns on all the decode strictness. But then, since it's layered, if people wanted to get picky, then they can pull it apart and do it themselves. We might be able to find similar opportunities here, or maybe we bundle it into the one suite of options--have options for all the different bits, but then one additional option "strict decode" that overrides them all. The API might get a bit too noisy and confusing though if there's a flat options space.

File: codec/dagcbor/multicodec.go [hunk 1]
```diff
@@ -28,7 +28,8 @@ func init() {
 // This is the function that will be registered in the default multicodec registry during package init time.
 func Decode(na datamodel.NodeAssembler, r io.Reader) error {
 	return DecodeOptions{
-		AllowLinks: true,
+		AllowLinks:  true,
+		MapSortMode: codec.MapSortMode_RFC7049,
```
[the remark this answers is not in the document]
Comment (mvdan) [reply codec/dagcbor/multicodec.go]: The default is now kept the same.

**Comment by mvdan** [comment]
I've rebased this and updated the change to keep the same default and make the option a general "be strict" boolean.

PTAL :)

**Comment by mvdan** [comment]
> @mvdan : will create the spec update PR?

Filed https://github.com/ipld/ipld/issues/196 so we don't forget. I haven't sent a PR as it's not a trivial change, and it's also not clear to me e.g. whether we want to apply the same thinking to other codecs.

**Comment by rvagg** [comment]
I've now got mixed feelings about this because it's so incomplete for something called "strict determinism".

Reviewing the list from the top of my head to see how deep this can go:

* floats must be 64 bits long
  - there may be a case for doing a round-trip on floats to ensure they use the canonical representation .. I'm not sure about this, it might be unnecessary but ieee754 has lots of weird spaces
* ints and negints must be small-as-possible
* all length-prefixed items must use small-as-possible ints - strings, bytes, maps, arrays, tag (and only 42 of course - must be `<tag><int 42>`)

Then there's all the little "you can't put that in there!" items, some of which refmt I think might reject already but we'd need to check:

* no novel major 7 (float) items, including whatever our latest position on `NaN` is (did we include it? if so, we should only allow the single correct form of it)
* no indefinite length items, and no `break` tokens

Most of this reaches down into refmt.. but you get the idea that having something called "strict determinism" that only does map key ordering and nothing else is a bit weak.

**Comment by mvdan** [comment]
This is definitely incomplete, but a first step adding the API. I see two options:

1) Merge the API, documenting that it is incomplete.
2) Merge the change but keep the API hidden until finished.

2 is the most correct option, as you point out, but I also worry that it might be one of those things in go-ipld-prime that doesn't quite get finished in a long time, for reasons like having to reach into refmt. So my personal take is that having something as an end user is better than nothing :)

Perhaps a middle ground is to call it `ExperimentalDeterminism`, documenting that it is incomplete and will eventually be replaced by `StrictDeterminism`.

**Comment by rvagg** [comment]
`ExperimentalDeterminism` - yes, I like this idea, sends the right message. Maybe link back to this thread in the docs even to note what still needs to be done.

**Comment by mvdan** [comment]
Done, PTAL

**Comment by mvdan** [comment]
I didn't add a link to this thread because the TODO already links to https://github.com/ipld/go-ipld-prime/pull/389

**Review by rvagg — APPROVED** [review]

**Pull request merged**